cs4660/graph/graph.py

```diff
@@ -0,0 +1,269 @@
+    def __init__(self, data):
+        self.data = data
+
+    def __str__(self):
+        return 'Node({})'.format(self.data)
+    def __repr__(self):
+        return 'Node({})'.format(self.data)
+
+    def __eq__(self, other_node):
+        return self.data == other_node.data
+    def __ne__(self, other):
+        return not self.__eq__(other)
+
+    def __hash__(self):
+        return hash(self.data)
+
+class Edge(object):
+    """Edge represents basic unit of graph connecting between two edges"""
+    def __init__(self, from_node, to_node, weight):
+        self.from_node = from_node
+        self.to_node = to_node
+        self.weight = weight
+    def __str__(self):
+        return 'Edge(from {}, to {}, weight {})'.format(self.from_node, self.to_node, self.weight)
+    def __repr__(self):
+        return 'Edge(from {}, to {}, weight {})'.format(self.from_node, self.to_node, self.weight)
+
+    def __eq__(self, other_node):
+        return self.from_node == other_node.from_node and self.to_node == other_node.to_node and self.weight == other_node.weight
+    def __ne__(self, other):
+        return not self.__eq__(other)
+
+    def __hash__(self):
+        return hash((self.from_node, self.to_node, self.weight))
+
+
+class AdjacencyList(object):
+    """
+    AdjacencyList is one of the graph representation which uses adjacency list to
+    store nodes and edges
+    """
+    def __init__(self):
+        # adjacencyList should be a dictonary of node to edges
+        self.adjacency_list = {}
+
+    def adjacent(self, node_1, node_2):
+        if node_1 in self.adjacency_list:
+            if node_2 in self.adjacency_list[node_1]:
+                 return True
+        return False
+
+    def neighbors(self, node):
+        rest = []
+        for edge in self.adjacency_list[node]:
+          rest.append(edge.to_node)
+        return rest
+        
+
+    def add_node(self, node):
+        if node in self.adjacency_list:
+           return False
+        else:
+            self.adjacency_list[node] = {}
+            return True
+
+
+    def remove_node(self, node):
+        if node in self.adjacency_list:
+            for from_node, edges in self.adjacency_list.items():
+                self.adjacency_list[from_node] = [node]
+                self.adjacency_list[node].remove(node)
+                return True
+
+        else:
+            return False
+
+    def add_edge(self, edge):
+        self.add_node(edge.from_node)
+        if(edge.to_node in self.adjacency_list[edge.from_node]):
+             return False
+        adj = self.adjacency_list[edge.from_node]     
+        adj.append(edge.to_node)
+        return True
+
+    def remove_edge(self, edge):
+         if edge.from_node in self.adjacency_list:
+             if edge.to_node in self.adjacency_list[edge.from_node]:
+                 self.adjacency_list[edge.from_node].remove(edge.to_node)
+                 return True
+         return False
```

Comment:
It seems like your code here are very inconsistent on the data structure used for the `adjacency_list`. Some method assumes the adjacency_list to be a dictionary of node to a list of nodes, some assumes the structure to be a dictionary of node to a list of edges while some assumes the dictionary to be node to a dictionary of nodes.